main.md

```diff
@@ -828,7 +830,112 @@ receipt of a `WWW-Authenticate: Bearer` challenge from a protected resource
 resource). The effect of this likely simplifies the logistics of phased 
 upgrades to protected resources in their support DPoP or even 
 prolonged deployments of protected resources with mixed token type support. 
-  
+
+# Authorization Server-Provided Nonce {#ASNonce}
+
+Including a nonce value contributed by the authorization server in the DPoP proof
+MAY used by authorization servers to limit the lifetime of DPoP proofs.
+The server is in control of when to require the use of a new nonce value
+in subsequent DPoP proofs.
+Without employing such a mechanism, a malicious party controlling the client
+(including potentially the end user)
+can create DPoP proofs for use arbitrarily far in the future.
+This section specifies how server-provided nonces are used with DPoP.
+
+An authorization server MAY supply a nonce value to be included by the client
```

Comment:
snippet from https://mailarchive.ietf.org/arch/msg/oauth/AiWXXnw5bVN4ePUVcbYFBVQGoRc/

Kind of a protocol compile error, if you will. The Authorization Server-Provided Nonce section uses the 'WWW-Authenticate: DPoP ... ' header in a 401 response to a token request to ask that a nonce be included in the DPoP proof in the subsequent request(s), which is sent as the value of the DPoP header. But based on the definition in this document and the underlying RFC7235, a 'WWW-Authenticate: DPoP' header is requesting that a 'Authorization: DPoP <access-token>' header be sent in the subsequent requests. The 'Authorization: DPoP <access-token>' header is only sent to protected resources. The way Authorization header is defined (can only have one in a request) and sometimes used for client secret basic auth are some of the reasons the DPoP proof is defined as separate header and not part of the Authorization header. Unfortunately, the result is that the error/challenge mechanics need to be slightly different between the AS and RS. I think the Authorization Server-Provided Nonce needs to be provided to the client in some other way. Perhaps as a parameter in the JSON of a token endpoint error response. Or in a header somehow. Or something else that I'm not thinking of. But the way it is in the PR right now doesn't semantically work.

Reply:
I think @danielfett 's suggestion of a new header that's applicable to all three situations (AS, RS and nonce) is probably the best path forward.

Reply:
The new `DPoP-Nonce` header suggestion works for me.  I'll plan to revise the PR accordingly.

Reply:
I wonder if there can be confusion around the DPoP Nonce vs. the OIDC Nonce. In particular, it would be bad if an implementer reads that the DPoP nonce must be the same as the ID token nonce (chosen by the client/RP). This would lead to a situation where an AS uses the more-or-less static value of the RP-selected nonce for the DPoP Nonce instead of a freshly chosen random one. 

I suggest to add at least a heads-up somewhere. Or we could call the claim for the nonce in the JWT 'dnc' or similar, to avoid confusion.

Reply:
I'd previously started writing a whole manifesto on why 'nonce' isn't the right term here. But decided to focus on other stuff, like the WWW-Authenticate, that seemed more important to address. But since you brought it up...

Reply:
The  registered nonce claim is a "Value used to associate a Client session with an ID Token" which is not what is going on in DPoP. But this isn't a hill I want to fight on. 

 https://www.iana.org/assignments/jwt/jwt.xhtml

Reply:
I realize that the registration language is more specific than our use.  However, our use does exactly match the definition in the IETF Security Glossary in RFC 4949, which says:

> nonce (I) A random or non-repeating value that is included in data exchanged by a protocol, usually for the purpose of guaranteeing liveness and thus detecting and protecting against replay attacks. (See: fresh.)

Reply:
Fair enough. Perhaps some kind of registry update is in order?

Reply:
I completely support the idea of a registry update.  I propose updating the description at https://www.iana.org/assignments/jwt/jwt.xhtml#claims from "Value used to associate a Client session with an ID Token" to "Value used to associate a Client session with an ID Token (MAY also be used for other kinds of nonces in other kinds of JWTs)".  I can do that as part of the OpenID Connect Core errata updates.